File: scripts/analysis/analyze_fraud_patterns.R
#!/usr/bin/env Rscript
# Analyze Fraud Patterns
# Identifies patterns and connections for administrative filings

library(dplyr)
library(stringr)
library(jsonlite)

# Configuration
DATA_DIR <- "../../data"
RESEARCH_DIR <- "../../research"
FILINGS_DIR <- "../../filings"
EVIDENCE_DIR <- "../../evidence"

# Load all relevant data
load_all_evidence <- function() {
  evidence <- list()

  # Load Skidmore firm data
  # Check for files in new location first, then fallback to root
  source_dir <- file.path(DATA_DIR, "source")
  firms_file <- if (file.exists(file.path(source_dir, "skidmore_all_firms_complete.csv"))) {
    file.path(source_dir, "skidmore_all_firms_complete.csv")
  } else if (file.exists("../../skidmore_all_firms_complete.csv")) {
    "../../skidmore_all_firms_complete.csv"
  } else {
    NULL
  }
  if (!is.null(firms_file) && file.exists(firms_file)) {
    evidence$firms <- read.csv(firms_file, stringsAsFactors = FALSE)
  }

  # Load DPOR connections
  connections_file <- file.path(DATA_DIR, "analysis", "dpor_skidmore_connections.csv")
  if (file.exists(connections_file)) {
    evidence$connections <- read.csv(connections_file, stringsAsFactors = FALSE)
  }

  # Load extracted PDF evidence
  pdf_evidence_file <- file.path(RESEARCH_EVIDENCE_DIR, "pdf_evidence_extracted.json")
  if (file.exists(pdf_evidence_file)) {
    evidence$pdf_data <- fromJSON(pdf_evidence_file)
  }

  return(evidence)
}

# Identify fraud indicators
identify_fraud_indicators <- function(evidence) {
  indicators <- list()

  # License-related fraud indicators
  if (!is.null(evidence$firms)) {
    # Check for license gaps
    indicators$license_gaps <- evidence$firms %>%
      filter(Gap.Years > 0) %>%
      select(Firm.Name, License.Number, Gap.Years, Notes)

    # Check for same address patterns (potential shell companies)
    address_counts <- evidence$firms %>%
      group_by(Address) %>%
      summarise(firm_count = n(), .groups = 'drop') %>%
      filter(firm_count > 1) %>%
      arrange(desc(firm_count))

    indicators$address_clusters <- address_counts
  }

  # Connection patterns
  if (!is.null(evidence$connections)) {
    # Multiple firms with same principal broker
    indicators$principal_broker_pattern <- evidence$connections %>%
      filter(connection_type == "Principal Broker") %>%
      group_by(state) %>%
      summarise(firm_count = n_distinct(firm_name), .groups = 'drop') %>%
      arrange(desc(firm_count))
  }

  # Timeline analysis
  if (!is.null(evidence$firms)) {
    # Firms licensed after Skidmore (potential backdating)
    indicators$timeline_issues <- evidence$firms %>%
      filter(!is.na(Initial.Cert.Date) & Initial.Cert.Date != "") %>%
      mutate(
        cert_date = as.Date(Initial.Cert.Date),
        skidmore_date = as.Date("2025-05-30"),
        licensed_after = cert_date > skidmore_date
      ) %>%
      filter(licensed_after) %>%
      select(Firm.Name, Initial.Cert.Date, Skidmore.License.Date)
  }

  return(indicators)
}

# Generate filing recommendations
generate_filing_recommendations <- function(indicators, evidence) {
  recommendations <- list()

  # Federal filings
  recommendations$federal <- list()

  if (nrow(indicators$license_gaps) > 0) {
    recommendations$federal$ftc <- list(
      reason = "Multiple firms with significant license gaps and same principal broker",
      evidence_count = nrow(indicators$license_gaps),
      priority = "High"
    )
  }

  if (!is.null(indicators$address_clusters) && nrow(indicators$address_clusters) > 0) {
    recommendations$federal$cfpb <- list(
      reason = "Address clustering suggests potential shell company scheme",
      evidence_count = nrow(indicators$address_clusters),
      priority = "High"
    )
  }

  # State filings
  recommendations$state <- list()

  if (!is.null(indicators$principal_broker_pattern) &&
      is.data.frame(indicators$principal_broker_pattern) &&
      nrow(indicators$principal_broker_pattern) > 0) {
    for (i in 1:nrow(indicators$principal_broker_pattern)) {
      state <- indicators$principal_broker_pattern$state[i]
      firm_count <- indicators$principal_broker_pattern$firm_count[i]

      recommendations$state[[state]] <- list(
        reason = paste("Multiple firms (", firm_count, ") listing same principal broker"),
        evidence_count = firm_count,
        priority = "High",
        filing_type = "License violation complaint"
      )
    }
  }

  # Local filings
  recommendations$local <- list(
    consumer_protection = list(
      reason = "Potential consumer fraud and deceptive business practices",
      priority = "Medium"
    )
  )

  return(recommendations)
}

# Create filing checklist
create_filing_checklist <- function(recommendations, evidence) {
  checklist <- data.frame(
    agency = character(),
    filing_type = character(),
    priority = character(),
    evidence_available = character(),
    status = character(),
    notes = character(),
    stringsAsFactors = FALSE
  )

  # Add federal filings
  if (length(recommendations$federal) > 0) {
    for (agency in names(recommendations$federal)) {
      rec <- recommendations$federal[[agency]]
      checklist <- rbind(checklist, data.frame(
        agency = toupper(agency),
        filing_type = "Federal complaint",
        priority = rec$priority,
        evidence_available = ifelse(is.null(rec$evidence_count), "Yes", paste(rec$evidence_count, "items")),
        status = "Pending",
        notes = rec$reason,
        stringsAsFactors = FALSE
      ))
    }
  }

  # Add state filings
  if (length(recommendations$state) > 0) {
    for (state in names(recommendations$state)) {
      rec <- recommendations$state[[state]]
      checklist <- rbind(checklist, data.frame(
        agency = paste("State:", state),
        filing_type = rec$filing_type,
        priority = rec$priority,
        evidence_available = paste(rec$evidence_count, "firms"),
        status = "Pending",
        notes = rec$reason,
        stringsAsFactors = FALSE
      ))
    }
  }

  return(checklist)
}

# Main analysis
main_analysis <- function() {
  cat("Loading evidence...\n")
  evidence <- load_all_evidence()

  cat("Identifying fraud indicators...\n")
  indicators <- identify_fraud_indicators(evidence)

  cat("Generating filing recommendations...\n")
  recommendations <- generate_filing_recommendations(indicators, evidence)

  cat("Creating filing checklist...\n")
  checklist <- create_filing_checklist(recommendations, evidence)

  # Save results
  dir.create(RESEARCH_DIR, showWarnings = FALSE, recursive = TRUE)
  dir.create(FILINGS_DIR, showWarnings = FALSE, recursive = TRUE)

  indicators_file <- file.path(RESEARCH_ANOMALIES_DIR, "fraud_indicators.json")
  write_json(indicators, indicators_file, pretty = TRUE)
  cat("Saved fraud indicators to:", indicators_file, "\n")

  recommendations_file <- file.path(RESEARCH_SUMMARIES_DIR, "filing_recommendations.json")
  write_json(recommendations, recommendations_file, pretty = TRUE)
  cat("Saved filing recommendations to:", recommendations_file, "\n")

  checklist_file <- file.path(FILINGS_DIR, "filing_checklist.csv")
  write.csv(checklist, checklist_file, row.names = FALSE)
  cat("Saved filing checklist to:", checklist_file, "\n")

  # Print summary
  cat("\n=== Fraud Pattern Analysis Summary ===\n")
  cat("License Gaps Found:", ifelse(is.null(indicators$license_gaps), 0, nrow(indicators$license_gaps)), "\n")
  cat("Address Clusters Found:", ifelse(is.null(indicators$address_clusters), 0, nrow(indicators$address_clusters)), "\n")
  cat("Federal Filings Recommended:", length(recommendations$federal), "\n")
  cat("State Filings Recommended:", length(recommendations$state), "\n")
  cat("\nFiling Checklist:\n")
  print(checklist)
}

# Run if executed as script
if (!interactive()) {
  main_analysis()
}

File: scripts/analysis/analyze_fraud_patterns.py
#!/usr/bin/env python3
"""
Analyze Fraud Patterns (Python)
Identifies patterns and connections for administrative filings
"""

import sys
import json
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Any
import pandas as pd

sys.path.insert(0, str(Path(__file__).parent.parent.parent))
from scripts.utils.paths import DATA_DIR, DATA_SOURCE_DIR, RESEARCH_DIR, FILINGS_DIR, RESEARCH_CONNECTIONS_DIR

def load_all_evidence() -> Dict[str, Any]:
    """Load all relevant data"""
    evidence = {}

    # Load Skidmore firm data
    firms_file = DATA_SOURCE_DIR / "skidmore_all_firms_complete.csv"
    if firms_file.exists():
        evidence['firms'] = pd.read_csv(firms_file)

    # Load DPOR connections
    connections_file = RESEARCH_CONNECTIONS_DIR / "dpor_skidmore_connections.csv"
    if connections_file.exists():
        evidence['connections'] = pd.read_csv(connections_file)

    # Load extracted PDF evidence
    pdf_evidence_file = RESEARCH_DIR / "pdf_evidence_extracted.json"
    if pdf_evidence_file.exists():
        with open(pdf_evidence_file, 'r') as f:
            evidence['pdf_data'] = json.load(f)

    return evidence

def identify_fraud_indicators(evidence: Dict[str, Any]) -> Dict[str, Any]:
    """Identify fraud indicators"""
    indicators = {}

    firms = evidence.get('firms', pd.DataFrame())
    if not firms.empty:
        # License gaps
        if 'Gap.Years' in firms.columns:
            indicators['license_gaps'] = firms[firms['Gap.Years'] > 0][
                ['Firm.Name', 'License.Number', 'Gap.Years', 'Notes']
            ].to_dict('records')

        # Address clusters (potential shell companies)
        if 'Address' in firms.columns:
            address_counts = firms.groupby('Address').size().reset_index(name='firm_count')
            indicators['address_clusters'] = address_counts[address_counts['firm_count'] > 1].sort_values('firm_count', ascending=False).to_dict('records')

    # Connection patterns
    connections = evidence.get('connections', pd.DataFrame())
    if not connections.empty and 'connection_type' in connections.columns:
        pb_pattern = connections[connections['connection_type'] == 'Principal Broker']
        if not pb_pattern.empty and 'state' in pb_pattern.columns:
            indicators['principal_broker_pattern'] = pb_pattern.groupby('state')['firm_name'].nunique().reset_index(name='firm_count').sort_values('firm_count', ascending=False).to_dict('records')

    # Timeline issues
    if not firms.empty and 'Initial.Cert.Date' in firms.columns:
        firms_with_dates = firms[firms['Initial.Cert.Date'].notna() & (firms['Initial.Cert.Date'] != '')]
        if not firms_with_dates.empty:
            firms_with_dates = firms_with_dates.copy()
            firms_with_dates['cert_date'] = pd.to_datetime(firms_with_dates['Initial.Cert.Date'], errors='coerce')
            firms_with_dates['skidmore_date'] = pd.Timestamp('2025-05-30')
            firms_with_dates['licensed_after'] = firms_with_dates['cert_date'] > firms_with_dates['skidmore_date']
            indicators['timeline_issues'] = firms_with_dates[firms_with_dates['licensed_after']][
                ['Firm.Name', 'Initial.Cert.Date', 'Skidmore.License.Date']
            ].to_dict('records')

    return indicators

def generate_filing_recommendations(indicators: Dict[str, Any], evidence: Dict[str, Any]) -> Dict[str, Any]:
    """Generate filing recommendations"""
    recommendations = {'federal': {}, 'state': {}, 'local': {}}

    # Federal filings
    if indicators.get('license_gaps'):
        recommendations['federal']['ftc'] = {
            'reason': 'Multiple firms with significant license gaps and same principal broker',
            'evidence_count': len(indicators['license_gaps']),
            'priority': 'High'
        }

    if indicators.get('address_clusters'):
        recommendations['federal']['cfpb'] = {
            'reason': 'Address clustering suggests potential shell company scheme',
            'evidence_count': len(indicators['address_clusters']),
            'priority': 'High'
        }

    # State filings
    if indicators.get('principal_broker_pattern'):
        for pattern in indicators['principal_broker_pattern']:
            state = pattern['state']
            firm_count = pattern['firm_count']
            recommendations['state'][state] = {
                'reason': f'Multiple firms ({firm_count}) listing same principal broker',
                'evidence_count': firm_count,
                'priority': 'High',
                'filing_type': 'License violation complaint'
            }

    # Local filings
    recommendations['local']['consumer_protection'] = {
        'reason': 'Potential consumer fraud and deceptive business practices',
        'priority': 'Medium'
    }

    return recommendations

def create_filing_checklist(recommendations: Dict[str, Any], evidence: Dict[str, Any]) -> pd.DataFrame:
    """Create filing checklist"""
    checklist_rows = []

    # Federal filings
    for agency, rec in recommendations.get('federal', {}).items():
        checklist_rows.append({
            'agency': agency.upper(),
            'filing_type': 'Federal complaint',
            'priority': rec['priority'],
            'evidence_available': f"{rec.get('evidence_count', 'N/A')} items",
            'status': 'Pending',
            'notes': rec['reason']
        })

    # State filings
    for state, rec in recommendations.get('state', {}).items():
        checklist_rows.append({
            'agency': f'State: {state}',
            'filing_type': rec.get('filing_type', 'License violation complaint'),
            'priority': rec['priority'],
            'evidence_available': f"{rec.get('evidence_count', 'N/A')} firms",
            'status': 'Pending',
            'notes': rec['reason']
        })

    return pd.DataFrame(checklist_rows)

def main_analysis():
    """Main analysis"""
    print("Loading evidence...")
    evidence = load_all_evidence()

    print("Identifying fraud indicators...")
    indicators = identify_fraud_indicators(evidence)

    print("Generating filing recommendations...")
    recommendations = generate_filing_recommendations(indicators, evidence)

    print("Creating filing checklist...")
    checklist = create_filing_checklist(recommendations, evidence)

    # Save results
    RESEARCH_DIR.mkdir(parents=True, exist_ok=True)
    FILINGS_DIR.mkdir(parents=True, exist_ok=True)

    indicators_file = RESEARCH_DIR / "fraud_indicators.json"
    with open(indicators_file, 'w') as f:
        json.dump(indicators, f, indent=2, default=str)
    print(f"Saved fraud indicators to: {indicators_file}")

    recommendations_file = RESEARCH_DIR / "filing_recommendations.json"
    with open(recommendations_file, 'w') as f:
        json.dump(recommendations, f, indent=2, default=str)
    print(f"Saved filing recommendations to: {recommendations_file}")

    checklist_file = FILINGS_DIR / "filing_checklist.csv"
    checklist.to_csv(checklist_file, index=False)
    print(f"Saved filing checklist to: {checklist_file}")

    # Print summary
    print("\n=== Fraud Pattern Analysis Summary ===")
    print(f"License Gaps Found: {len(indicators.get('license_gaps', []))}")
    print(f"Address Clusters Found: {len(indicators.get('address_clusters', []))}")
    print(f"Federal Filings Recommended: {len(recommendations.get('federal', {}))}")
    print(f"State Filings Recommended: {len(recommendations.get('state', {}))}")
    print("\nFiling Checklist:")
    print(checklist.to_string(index=False))

if __name__ == "__main__":
    main_analysis()
